modules/openapi-generator/src/main/java/org/openapitools/codegen/InlineModelResolver.java

```diff
@@ -104,95 +105,220 @@ private void flattenPaths(OpenAPI openAPI) {
-                        openAPI.getComponents().getRequestBodies().put(modelName, rb);
+                        newAllOf.add(inner);
                     }
-
-                    // update requestBody to use $ref instead of inline def
-                    requestBody.set$ref(modelName);
-
                 }
+                m.setAllOf(newAllOf);
             }
-        } else if (model instanceof ArraySchema) {
-            ArraySchema am = (ArraySchema) model;
-            Schema inner = am.getItems();
-            if (inner instanceof ObjectSchema) {
-                ObjectSchema op = (ObjectSchema) inner;
-                if (op.getProperties() != null && op.getProperties().size() > 0) {
-                    flattenProperties(openAPI, op.getProperties(), pathname);
-                    // Generate a unique model name based on the title.
-                    String modelName = resolveModelName(op.getTitle(), null);
-                    Schema innerModel = modelFromProperty(openAPI, op, modelName);
-                    String existing = matchGenerated(innerModel);
-                    if (existing != null) {
-                        Schema schema = new Schema().$ref(existing);
-                        schema.setRequired(op.getRequired());
-                        am.setItems(schema);
+            if (m.getAnyOf() != null) {
+                List<Schema> newAnyOf = new ArrayList<Schema>();
+                for (Schema inner : m.getAnyOf()) {
+                    // Recurse to create $refs for inner models
+                    gatherInlineModels(inner, modelPrefix + "AnyOf");
+                    if (isModelNeeded(inner)) {
+                        Schema refSchema = this.makeSchemaResolve(modelPrefix, "AnyOf", inner);
+                        newAnyOf.add(refSchema); // replace with ref
                     } else {
-                        Schema schema = new Schema().$ref(modelName);
-                        schema.setRequired(op.getRequired());
-                        am.setItems(schema);
-                        addGenerated(modelName, innerModel);
-                        openAPI.getComponents().addSchemas(modelName, innerModel);
+                        newAnyOf.add(inner);
                     }
                 }
+                m.setAnyOf(newAnyOf);
+            }
+            if (m.getOneOf() != null) {
+                List<Schema> newOneOf = new ArrayList<Schema>();
+                for (Schema inner : m.getOneOf()) {
+                    // Recurse to create $refs for inner models
+                    gatherInlineModels(inner, modelPrefix + "OneOf");
+                    if (isModelNeeded(inner)) {
+                        Schema refSchema = this.makeSchemaResolve(modelPrefix, "OneOf", inner);
+                        newOneOf.add(refSchema); // replace with ref
+                    } else {
+                        newOneOf.add(inner);
+                    }
+                }
+                m.setOneOf(newOneOf);
```

Comment:
This looks good. If you wanted you could shorten this by:
- making a list of allOf, anyOf, and oneOf values.
Loop though the list of lists and handle each case accordingly.
Collapse this similar code from being in 3 places to one place.